### PR TITLE
Enable Cute Unit tests

### PR DESCRIPTION
**Pull request by AD2605**
Enables Cute unit tests. 
Currently draft because this is waiting on work-group static local memory extension implementation in dpcpp.

It compiles, but I am yet to run any tests, but would appreciate any initial feedback.

PS - Please ignore the commit history it shows for now, I am trying to fix that

### CODE REVIEW

File: test/unit/CMakeLists.txt [hunk 1]
```diff
@@ -169,6 +169,4 @@ foreach(SUBDIR ${SUBDIRS})
   add_subdirectory(${SUBDIR})
   add_dependencies(cutlass_test_unit cutlass_test_unit_${SUBDIR})
   add_dependencies(test_unit test_unit_${SUBDIR})
-
```
Comment (mehdi-goli) [inline test/unit/CMakeLists.txt:172]: these are noise in the PR

File: test/unit/CMakeLists.txt [hunk 2]
```diff
@@ -169,6 +169,4 @@ foreach(SUBDIR ${SUBDIRS})
...
 endforeach()
-
```
Comment (mehdi-goli) [inline test/unit/CMakeLists.txt:174]: same here

File: test/unit/cute/ampere/cp_async.cu [hunk 3]
```diff
@@ -38,65 +38,86 @@
...
+  using namespace syclcompat;
+#endif
+
+CUTLASS_GLOBAL void
 test(double const* g_in, double* g_out)
 {
-  extern __shared__ double smem[];
-
-  smem[threadIdx.x] = g_in[threadIdx.x];
+  #ifdef CUTLASS_ENABLE_SYCL
```
Comment (mehdi-goli) [inline test/unit/cute/ampere/cp_async.cu:58]:
```suggestion
  #ifdef __SYCL_DEVICE_ONLY__
```

File: test/unit/cute/ampere/cp_async.cu [hunk 4]
```diff
@@ -38,65 +38,86 @@
...
 }
 
-__global__ void
+CUTLASS_GLOBAL void
 test2(double const* g_in, double* g_out)
 {
   using namespace cute;
 
-  extern __shared__ double smem[];
+  #ifdef CUTLASS_ENABLE_SYCL
```
Comment (mehdi-goli) [inline test/unit/cute/ampere/cp_async.cu:75]:
```suggestion
  #ifdef __SYCL_DEVICE_ONLY__
```

File: test/unit/cute/ampere/ldsm.cu [hunk 5]
```diff
@@ -43,21 +43,32 @@
...
 ldsm_test_device(uint16_t* g_in, uint16_t* g_out)
 {
   constexpr int count = sizeof(T) / 4;
-  int tid = threadIdx.x;
-  int stride = blockDim.x;
+  int tid = threadIdxX();
+  int stride = blockDimX();
 
   // load input gmem -> smem
+  #if defined(CUTLASS_ENABLE_SYCL)
```
Comment (mehdi-goli) [inline test/unit/cute/ampere/ldsm.cu:60]:
```suggestion
  #if defined(CUTLASS_ENABLE_SYCL)
```

File: test/unit/cute/ampere/ldsm.cu [hunk 6]
```diff
@@ -76,13 +87,18 @@ ldsm_test_device(uint16_t* g_in, uint16_t* g_out)
...
 
 template <class TiledCopy, class SmemLayout>
-__global__ void
+CUTLASS_GLOBAL void
 ldsm_test_device_cute(uint16_t* g_in, uint16_t* g_out,
                       TiledCopy tiled_copy, SmemLayout smem_layout)
 {
   using namespace cute;
-
+  #if defined(CUTLASS_ENABLE_SYCL)
```
Comment (mehdi-goli) [inline test/unit/cute/ampere/ldsm.cu:95]:
```suggestion
  #if defined(__SYCL_DEVICE_ONLY__)
```

File: test/unit/cute/ampere/ldsm.cu [hunk 7]
```diff
@@ -125,22 +141,26 @@ TEST(SM80_CuTe_Ampere, Ldsm)
...
   //
 
   {
-  thrust::device_vector<uint16_t> d_out(count);
-  ldsm_test_device<uint32_t><<<1, 32>>>(
-    thrust::raw_pointer_cast(d_in.data()),
-    thrust::raw_pointer_cast(d_out.data()));
-  thrust::host_vector<uint16_t> h_out = d_out;
+  device_vector<uint16_t> d_out(count);
+  #if defined(CUTLASS_ENABLE_SYCL)
```
Comment (mehdi-goli) [inline test/unit/cute/ampere/ldsm.cu:156]:
```suggestion
  #if defined(__SYCL_DEVICE_ONLY__)
```

File: test/unit/cute/hopper/bulk_load.cu [hunk 8]
```diff
@@ -52,14 +52,18 @@ struct SharedStorage {
...
 template <class T, class GmemLayout, class SmemLayout>
-__global__ void
+CUTLASS_GLOBAL void
 bulk_copy_test_device_cute(T const* g_in,
                            T      * g_out,
                            GmemLayout gmem_layout,
                            SmemLayout smem_layout)
 {
   // Use Shared Storage structure to allocate and distribute aligned SMEM addresses
+  #if defined(CUTLASS_ENABLE_SYCL)
```
Comment (mehdi-goli) [inline test/unit/cute/hopper/bulk_load.cu:62]:
```suggestion
  #if defined(__SYCL_DEVICE_ONLY__)
```

File: test/unit/cute/hopper/bulk_store.cu [hunk 9]
```diff
@@ -51,14 +51,18 @@ struct SharedStorage {
...
 template <class T, class GmemLayout, class SmemLayout>
-__global__ void
+CUTLASS_GLOBAL void
 bulk_copy_test_device_cute(T const* g_in,
                            T      * g_out,
                            GmemLayout gmem_layout,
                            SmemLayout smem_layout)
 {
   // Use Shared Storage structure to allocate and distribute aligned SMEM addresses
+  #if defined(CUTLASS_ENABLE_SYCL)
```
Comment (mehdi-goli) [inline test/unit/cute/hopper/bulk_store.cu:61]:
```suggestion
  #if defined(__SYCL_DEVICE_ONLY__)
```

File: test/unit/cute/hopper/stsm.cu [hunk 10]
```diff
@@ -42,27 +42,31 @@
...
+  int stride = blockDimX();
 
   // load input gmem -> rmem
   uint32_t reg[count];
   for (int i = 0; i < (sizeof(T) / 4); i++) {
     reg[i] = reinterpret_cast<uint32_t*>(g_in)[tid + (stride * i)];
   }
-
+  
+  #if defined(CUTLASS_ENABLE_SYCL)
```
Comment (mehdi-goli) [inline test/unit/cute/hopper/stsm.cu:58]:
```suggestion
  #if defined(__SYCL_DEVICE_ONLY__)
```

File: test/unit/cute/hopper/stsm.cu [hunk 11]
```diff
@@ -71,19 +75,23 @@ stsm_test_device(uint16_t* g_in, uint16_t* g_out)
...
 
 template <class TiledCopy, class SmemLayout>
-__global__ void
+CUTLASS_GLOBAL void
 stsm_test_device_cute(uint16_t* g_in, uint16_t* g_out,
                       TiledCopy tiled_copy, SmemLayout smem_layout)
 {
   using namespace cute;
 
+  #if defined(CUTLASS_ENABLE_SYCL)
```
Comment (mehdi-goli) [inline test/unit/cute/hopper/stsm.cu:84]:
```suggestion
  #if defined(__SYCL_DEVICE_ONLY__)
```

File: test/unit/cute/hopper/tma_load_testbed.hpp [hunk 12]
```diff
@@ -62,7 +62,11 @@ tma_test_device_cute(T const* g_in, T* g_out,
   CUTE_STATIC_ASSERT_V(product_each(shape(cta_tiler)) == product_each(shape(smem_layout)));
 
   // Use Shared Storage structure to allocate and distribute aligned SMEM addresses
+  #if defined(CUTLASS_ENABLE_SYCL)
```
Comment (mehdi-goli) [inline test/unit/cute/hopper/tma_load_testbed.hpp:65]:
```suggestion
  #if defined(__SYCL_DEVICE_ONLY__)
```

File: test/unit/cute/ampere/cooperative_copy.cu [hunk 13]
```diff
@@ -73,25 +77,29 @@ cooperative_copy_default_gs(T const* g_in, T* g_out, GMemLayout const& gmem_layo
...
 }
 
 // ss --> shared to shared
 template <int MaxVecBits, uint32_t ThreadBlockSize, class T, class Layout1, class Layout2>
-__device__ void
+CUTLASS_DEVICE void
 cooperative_copy_default_ss(T const* g_in, T* g_out, Layout1 const& layout1, Layout2 const& layout2)
 {
   using namespace cute;
+  #if defined(CUTLASS_ENABLE_SYCL)
```
Comment (mehdi-goli) [inline test/unit/cute/ampere/cooperative_copy.cu:98]:
```suggestion
  #if defined(__SYCL_DEVICE_ONLY__)
```

File: test/unit/cute/CMakeLists.txt [hunk 14]
```diff
@@ -63,15 +63,15 @@ else()
   
   add_subdirectory(layout)
   add_subdirectory(core)
+  add_subdirectory(msvc_compilation)
```
Comment (AD2605) [inline test/unit/cute/CMakeLists.txt:66]: @aacostadiaz , I had not added `msvc_compilation` on purpose, as we do not use the Microsoft visual C++ compiler,
is this still required ?

**Review by aacostadiaz — APPROVED** [review]

**Review by mehdi-goli — APPROVED** [review]

**Pull request merged**